x/smart/psiphon_dialer.go

```diff
@@ -1,23 +1,38 @@
+//go:build psiphon
 // +build psiphon
+
 // If the build tag `psiphon` is set, allow importing and calling psiphon
 
 package smart
 
 import (
 	"context"
-	"errors"
 	"fmt"
+	"os"
+	"path"
 
 	"github.com/Jigsaw-Code/outline-sdk/transport"
 	"github.com/Jigsaw-Code/outline-sdk/x/psiphon"
 )
 
-func newPsiphonDialer(ctx context.Context, psiphonJSON []byte) (transport.StreamDialer, error) {
-	// Test calling psiphon package to prove it works
-	dialer := psiphon.GetSingletonDialer()
-	fmt.Printf("Initializing psiphon dialer: %+v\n", dialer)
+func newPsiphonDialer(finder *StrategyFinder, ctx context.Context, psiphonJSON []byte) (transport.StreamDialer, error) {
+	config := &psiphon.DialerConfig{ProviderConfig: psiphonJSON}
+
+	cacheBaseDir, err := os.UserCacheDir()
+	if err != nil {
+		return nil, fmt.Errorf("Failed to get the user cache directory: %w", err)
+	}
 
-	//TODO(laplante): set up and use psiphon dialer
+	config.DataRootDirectory = path.Join(cacheBaseDir, "psiphon")
+	if err := os.MkdirAll(config.DataRootDirectory, 0700); err != nil {
+		return nil, fmt.Errorf("Failed to create storage directory: %w", err)
+	}
+	finder.logCtx(ctx, "Using data store in %v\n", config.DataRootDirectory)
+
+	dialer := psiphon.GetSingletonDialer()
+	if err := dialer.Start(ctx, config); err != nil {
```

Comment:
This is not the correct Context to pass to Start. This context is for the creation of the Dialer. If the user imposes a deadline of 5 seconds to create the dialer, it shouldn't get cancelled after it returns, but it will with the current code.

We need a new context that propagates the original context cancellation, but only during the construction.

Reply:
Perhaps use https://pkg.go.dev/context#AfterFunc on the original context to cancel the Start context.

Reply:
Merged https://github.com/Jigsaw-Code/outline-sdk/pull/427

And switched the context passed into this file to be raceCtx, so we stop trying/logging after the race has ended.